File: tests/ssurjano_benchmark/detpep108d.py
import numpy as np

def detpep108d(xx):
    xx = xx.reshape(-1, 4) # reshape input array to 2D with shape (n, 4)
    
    term1 = 4 * ((xx[0] - 2 + 8*xx[1] - 8*np.square(xx[1])).reshape(-1, 1)**2)
    term2 = np.power((3-4*xx[1]), 2)
    term3 = 16 * np.sqrt(xx[2]+1) * (2*xx[2]-1)**2
    
    xxmat = np.tile(xx[2:], (6, 6)) # create 8x8 matrix with zeros in first row and column
    xxmatlow = xxmat[:5, :5] # select lower triangle of matrix
    inner = np.sum(xxmatlow, axis=1) # sum along rows to get inner product
    outer = np.sum(inner*np.arange(4)) # multiply by indices and sum
    
    y = term1 + term2 + term3 + outer
    return(y)

File: tests/ssurjano_benchmark/detpep108d.r
detpep108d <- function(xx)
{
  ##########################################################################
  #
  # DETTE & PEPELYSHEV (2010) 8-DIMENSIONAL FUNCTION
  #
  # Authors: Sonja Surjanovic, Simon Fraser University
  #          Derek Bingham, Simon Fraser University
  # Questions/Comments: Please email Derek Bingham at dbingham@stat.sfu.ca.
  #
  # Copyright 2013. Derek Bingham, Simon Fraser University.
  #
  # THERE IS NO WARRANTY, EXPRESS OR IMPLIED. WE DO NOT ASSUME ANY LIABILITY
  # FOR THE USE OF THIS SOFTWARE.  If software is modified to produce
  # derivative works, such modified software should be clearly marked.
  # Additionally, this program is free software; you can redistribute it 
  # and/or modify it under the terms of the GNU General Public License as 
  # published by the Free Software Foundation; version 2.0 of the License. 
  # Accordingly, this program is distributed in the hope that it will be 
  # useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
  # of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU 
  # General Public License for more details.
  #
  # For function details and reference information, see:
  # http://www.sfu.ca/~ssurjano/
  #
  ##########################################################################
  #
  # INPUT:
  #
  # xx = c(x1, x2, ..., x8)
  #
  #########################################################################
  
  x1 <- xx[1]
  x2 <- xx[2]
  x3 <- xx[3]
  ii <- c(4:8)
  
  term1 <- 4 * (x1 - 2 + 8*x2 - 8*x2^2)^2
  term2 <- (3 - 4*x2)^2
  term3 <- 16 * sqrt(x3+1) * (2*x3-1)^2
  
  xxmat <- matrix(rep(xx[3:8],times=6), 6, 6, byrow=TRUE)
  xxmatlow <- xxmat
  xxmatlow[upper.tri(xxmatlow)] <- 0
  
  inner <- rowSums(xxmatlow)
  inner <- inner[2:6]
  outer <- sum(ii*log(1+inner))
  
  y <- term1 + term2 + term3 + outer
  return(y)
}
